examples/model_compress/sparse_learning_example/main.py

```diff
@@ -0,0 +1,280 @@
+            print_and_log('Growth mode: {0}'.format(args.growth))
+            print_and_log('Redistribution mode: {0}'.format(args.redistribution))
+            print_and_log('='*60)
+
+        # add custom prune/growth/redisribution here
+        if args.prune == 'magnitude_variance':
+            print('Using magnitude-variance pruning. Switching to Adam optimizer...')
+            args.prune = magnitude_variance_pruning
+            args.optimizer = 'adam'
+        if args.redistribution == 'variance':
+            print('Using variance redistribution. Switching to Adam optimizer...')
+            args.redistribution = variance_redistribution
+            args.optimizer = 'adam'
+
+        optimizer = None
+        if args.optimizer == 'sgd':
+            optimizer = optim.SGD(model.parameters(),lr=args.lr,momentum=args.momentum,weight_decay=args.l2, nesterov=True)
+        elif args.optimizer == 'adam':
+            optimizer = optim.Adam(model.parameters(),lr=args.lr,weight_decay=args.l2)
+        else:
+            print('Unknown optimizer: {0}'.format(args.optimizer))
+            raise Exception('Unknown optimizer.')
+
+        lr_scheduler = optim.lr_scheduler.StepLR(optimizer, args.decay_frequency, gamma=0.1)
+
+        if args.resume:
+            if os.path.isfile(args.resume):
+                print_and_log("=> loading checkpoint '{}'".format(args.resume))
+                checkpoint = torch.load(args.resume)
+                args.start_epoch = checkpoint['epoch']
+                model.load_state_dict(checkpoint['state_dict'])
+                optimizer.load_state_dict(checkpoint['optimizer'])
+                print_and_log("=> loaded checkpoint '{}' (epoch {})"
+                      .format(args.resume, checkpoint['epoch']))
+                print_and_log('Testing...')
+                evaluate(args, model, device, test_loader)
+                model.feats = []
+                model.densities = []
+                plot_class_feature_histograms(args, model, device, train_loader, optimizer)
+            else:
+                print_and_log("=> no checkpoint found at '{}'".format(args.resume))
+
+
+        if args.fp16:
+            print('FP16')
+            optimizer = FP16_Optimizer(optimizer,
+                                       static_loss_scale = None,
+                                       dynamic_loss_scale = True,
+                                       dynamic_loss_args = {'init_scale': 2 ** 16})
+            model = model.half()
+
+        pruner = None
+        if not args.dense:
+            decay = CosineDecay(args.prune_rate, len(train_loader)*(args.epochs))
+            #decay = LinearDecay(args.prune_rate, len(train_loader)*(args.epochs))
+            pruner = SparsePruner(optimizer, decay, density=args.density, sparse_init='constant', prune_rate=args.prune_rate, prune_mode=args.prune, growth_mode=args.growth, redistribution_mode=args.redistribution,
```

Comment:
suggest to use comment to highlight that this is NNI related code to leverage the pruner in NNI